include/wil/registry.h

```diff
@@ -17,16 +17,2855 @@
+         */
+        inline HRESULT get_value_dword_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _Out_ DWORD* return_value) WI_NOEXCEPT
+        {
+            return ::wil::reg::get_value_nothrow(key, nullptr, value_name, return_value);
+        }
+
+        /**
+         * \brief Reads a REG_QWORD value under a specified key
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param subkey A string to append to the HKEY to attempt to read from
+         *        can be nullptr if not needed
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \param[out] return_value A DWORD64 receiving the value read from the registry
+         * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
+         */
+        inline HRESULT get_value_qword_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, _Out_ DWORD64* return_value) WI_NOEXCEPT
+        {
+            return ::wil::reg::get_value_nothrow(key, subkey, value_name, return_value);
+        }
+
+        /**
+         * \brief Reads a REG_QWORD value under a specified key
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \param[out] return_value A DWORD64 receiving the value read from the registry
+         * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
+         */
+        inline HRESULT get_value_qword_nothrow(HKEY key, _In_opt_ PCWSTR value_name, _Out_ DWORD64* return_value) WI_NOEXCEPT
+        {
+            return ::wil::reg::get_value_nothrow(key, nullptr, value_name, return_value);
+        }
+
+#if defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
+        /**
+         * \brief Reads a REG_SZ value under a specified key
+         * \param key An opened registry key, or fixed registry key as the base key, from which to append the path
+         * \param subkey A string to append to the HKEY to attempt to read from
+         *        can be nullptr if not needed
+         * \param value_name A string specifying the name of the registry value to read from
+         *        can be nullptr to read from the unnamed default registry value
+         * \param[out] return_value A std::wstring receiving the value read from the registry
+         * \return HRESULT error code indicating success or failure (does not throw C++ exceptions)
+         */
+        inline HRESULT get_value_nothrow(HKEY key, _In_opt_ PCWSTR subkey, _In_opt_ PCWSTR value_name, ::std::wstring& return_value) WI_NOEXCEPT

```

Comment:
It's a little wacky that a nothrow marked thing uses a C++ thing that by-definition throws on failure to allocate.  Consider removing this nothrow - if you want nothrow, you have to use a non-C++ string type.

Reply:
(Applies elsewhere.)

Reply:
the idea is that often folks want/need a no-throw function specifically at their callsite (even though clearly if using a std::wstring then they must support C++ exceptions in general). it's not completely unheard of; there are functions even in the STL that can write to a container passed by ref, but is no-throw. e.g. std::swap.

Reply:
happy to discuss this more if you want to follow-up